query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala

```diff
@@ -395,75 +397,283 @@ final case class LabelCardinalityExec(queryContext: QueryContext,
+
     val rvs = source match {
       case tsMemStore: TimeSeriesMemStore =>
         Observable.eval {
-          val topkCards = tsMemStore.topKCardinality(dataset, Seq(shard), shardKeyPrefix, k, addInactive)
-          // Include each into an ItemSketch.
-          val sketch = new ItemsSketch[String](TopkCardExec.MAX_ITEMSKETCH_MAP_SIZE)
-          topkCards.foreach{ card =>
-            val count = if (addInactive) card.tsCount.toLong else card.activeTsCount.toLong
-            sketch.update(card.childName, count)
+          val countMap = if (groupDepth == shardKeyPrefix.size - 1) {
+            // shardKeyPrefix is as deep as the groupDepth; we execute only a single topKCardinality
+            //   call on the "prefix's prefix", then map the prefix to its cardinality iff the
+            //   "deepest" prefix label is found in the results.
+            val res = tsMemStore.topKCardinality(dataset, Seq(shard),
+                                                 shardKeyPrefix.dropRight(1),
+                                                 MAX_RESPONSE_SIZE, ADD_INACTIVE)
+              .find(card => card.childName == shardKeyPrefix.last)
+            if (res.nonEmpty) {
+              Map(shardKeyPrefix.map(_.utf8).toSeq -> CardCounts(res.get.activeTsCount, res.get.tsCount))
+            } else {
+              Map()
+            }
+          } else {
+            // groupDepth is deeper than shardKeyPrefix.
+            // We will "expand" shardKeyPrefix into all of its child prefixes at groupDepth - 1,
+            //   then we'll pass the children to memstore.topKCardinality and map the results.
+
+            // The conditional exists because PrefixIterator currently
+            //   needs to extend a prefix to a size > 0
+            val prefixIter = if (groupDepth > 0) {
+              // Note: groupDepth equals the length of the prefix to extend to (at groupDepth - 1)
+              new PrefixIterator(tsMemStore, shardKeyPrefix, groupDepth)
+            } else {
+              Seq(Seq()).iterator
+            }
+
+            prefixIter.flatMap { prefix =>
+              tsMemStore.topKCardinality(dataset, Seq(shard), prefix,
+                                         MAX_RESPONSE_SIZE, ADD_INACTIVE)
+                .map{ card =>
+                  // Concatenate `prefix` with the least-significant name.
+                  val totalPrefix = Seq(prefix, Seq(card.childName)).flatMap(_.map(_.utf8)).toSeq
+                  totalPrefix -> CardCounts(card.activeTsCount, card.tsCount)
+                }
+            }.toMap
           }
-          // serialize the sketch; pack it into a RangeVector
-          val serSketch = ZeroCopyUTF8String(sketch.toByteArray(new ArrayOfStringsSerDe))
-          val it = Seq(SingleValueRowReader(serSketch)).iterator
+
+          // serialize the map; pack it into a RangeVector
+          val serMap = ZeroCopyUTF8String(serializeMap(countMap.toMap))
```

Comment:
The maximum string length that can be added is 65536.  See RecordBuilder.scala line 182. 
Can you think of adding each entry of the map as one record ?


cc @amolnayak311  Please check if your PR has this limitation.

Reply:
Alex, consider returning N rows of this schema: (group, activeInt, totalInt)
where group is a concatenation of strings with a delimiter.

Reply:
Updated: presenter removed, serialization removed.